File: 200-number-of-islands/number-of-islands.java
class Solution {
    void DFS(char[][] grid, int i, int j, int r, int c)
    {
        if(i==r || j==c || i<0 || j<0 || grid[i][j] == '0')
            return;
        grid[i][j] = '0';
        DFS(grid, i, j+1, r, c);
        DFS(grid, i-1, j, r, c);
        DFS(grid, i, j-1, r, c);
        DFS(grid, i+1, j, r, c);
    }
    public int numIslands(char[][] grid) {
       int r = grid.length;
       int c = grid[0].length;
       int ans = 0;
       for(int i=0; i<r; i++)
       {
         for(int j=0; j<c; j++)
         {
            if(grid[i][j] == '1')
            {
                DFS(grid, i, j, r, c);
                ans++;
            }
         }
       } 
       return ans;
    }
}

File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void BFS(vector<vector<char>>& grid, int i, int j, int n, int m){
        queue<pair<int, int>> q;
        q.push({i,j});
        vector<pair<int, int>> directions = {{0,1}, {1,0}, {0,-1}, {-1, 0}};
        while(!q.empty()){
            auto node = q.front();
            q.pop();
            for(auto [dx, dy] : directions){
                int newX = node.first + dx;
                int newY = node.second + dy;
                if(newX>=0 && newX<n && newY>=0 && newY<m && grid[newX][newY]=='1'){
                    grid[newX][newY] = '0';
                    q.push({newX, newY});
                }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        int ans = 0;
        for(int i=0; i<n; i++){
            for(int j=0; j<m; j++){
                if(grid[i][j]=='1'){
                    BFS(grid, i, j, n, m);
                    ans++;
                }
            }
        }
        return ans;
    }
};